java/tests/src/test/java/org/signal/libsignal/metadata/SealedSessionCipherTest.java

```diff
@@ -224,6 +228,44 @@ public void testProtocolException() throws UntrustedIdentityException, InvalidKe
     }
   }
 
+  public void testDecryptionErrorMessage() throws InvalidCertificateException, InvalidKeyException, InvalidMetadataMessageException, InvalidMetadataVersionException, ProtocolDuplicateMessageException, ProtocolInvalidKeyException, ProtocolInvalidKeyIdException, ProtocolInvalidMessageException, ProtocolInvalidVersionException, ProtocolLegacyMessageException, ProtocolNoSessionException, ProtocolUntrustedIdentityException, SelfSendException, UntrustedIdentityException {
+    TestInMemorySignalProtocolStore aliceStore = new TestInMemorySignalProtocolStore();
+    TestInMemorySignalProtocolStore bobStore   = new TestInMemorySignalProtocolStore();
+    SignalProtocolAddress bobAddress           = new SignalProtocolAddress("+14152222222", 1);
+
+    initializeSessions(aliceStore, bobStore, bobAddress);
+
+    ECKeyPair            trustRoot            = Curve.generateKeyPair();
+    CertificateValidator certificateValidator = new CertificateValidator(trustRoot.getPublicKey());
+    SenderCertificate    senderCertificate    = createCertificateFor(trustRoot, UUID.fromString("9d0652a3-dcc3-4d11-975f-74d61598733f"), "+14151111111", 1, aliceStore.getIdentityKeyPair().getPublicKey().getPublicKey(), 31337);
+    SealedSessionCipher  aliceCipher          = new SealedSessionCipher(aliceStore, UUID.fromString("9d0652a3-dcc3-4d11-975f-74d61598733f"), "+14151111111", 1);
+
+    // Send a message from Alice to Bob to set up the session.
+    byte[] ciphertext = aliceCipher.encrypt(bobAddress, senderCertificate, "smert za smert".getBytes());
+
+    SealedSessionCipher bobCipher = new SealedSessionCipher(bobStore, UUID.fromString("e80f7bbe-5b94-471e-bd8c-2173654ea3d1"), "+14152222222", 1);
+
+    bobCipher.decrypt(certificateValidator, ciphertext, 31335);
+
+    // Pretend Bob's reply fails to decrypt.
+    SignalProtocolAddress aliceAddress = new SignalProtocolAddress("9d0652a3-dcc3-4d11-975f-74d61598733f", 1);
+    SessionCipher bobUnsealedCipher = new SessionCipher(bobStore, aliceAddress);
+    CiphertextMessage bobMessage = bobUnsealedCipher.encrypt("reply".getBytes());
+
+    DecryptionErrorMessage errorMessage = DecryptionErrorMessage.forOriginalMessage(bobMessage.serialize(), bobMessage.getType(), 408);
+    PlaintextContent errorMessageContent = new PlaintextContent(errorMessage);
+    UnidentifiedSenderMessageContent errorMessageUsmc = new UnidentifiedSenderMessageContent(errorMessageContent, senderCertificate, UnidentifiedSenderMessageContent.CONTENT_HINT_SUPPLEMENTARY, Optional.<byte[]>absent());
+    byte[] errorMessageCiphertext = aliceCipher.encrypt(bobAddress, errorMessageUsmc);
+
+    DecryptionResult result = bobCipher.decrypt(certificateValidator, errorMessageCiphertext, 31335);
+    PlaintextContent bobErrorContent = result.getPlaintextContent().get();
+    DecryptionErrorMessage bobErrorMessage = bobErrorContent.getDecryptionErrorMessage().get();
```

Comment:
It seems that `PlaintextContent` may represent an error and it may not? But I cannot see a test for when it does not.

Reply:
Right now the only thing it ever represents is the error, but we might add more plaintext messages in the future. This parallels the existing Content proto used by the apps.

https://github.com/signalapp/Signal-Android/blob/fb1b46b67e791cde9c49e30e8adced3e997bc63e/libsignal/service/src/main/proto/SignalService.proto#L35-L42